### PR TITLE
Extract tag class out of pytato

### PR DESCRIPTION
**Pull request by nchristensen**

### CODE REVIEW

File: pytools/tag.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
+from __future__ import annotations
+from dataclasses import dataclass
+from typing import (Tuple, Any, FrozenSet)
```
Comment (inducer) [inline pytools/tag.py:3]:
```suggestion
from typing import Tuple, Any, FrozenSet
```

File: pytools/tag.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
+from __future__ import annotations
```
Comment (inducer) [inline pytools/tag.py:1]: This is not consistent with the Py3.6+ promise in `setup.py` (which I'd like to retain)

Comment (inducer) [inline pytools/tag.py:1]: This should be wired into the documentation, i.e.

- Get a module docstring explaining what it's for and including `.. auto*` directives for the parts being exposed.
- Have `.. automodule` for that included in the package docs.

Comment (inducer) [inline pytools/tag.py:1]: Needs a copyright/license header.

File: pytools/tag.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            raise ValueError(f"some name parts of {'.'.join(name_parts)} "
+                             "start with double underscores")
+        return cls(name_parts)
+
+
+# }}}
+
+# {{{ tag
+
+tag_dataclass = dataclass(init=True, eq=True, frozen=True, repr=True)
```
Comment (inducer) [inline pytools/tag.py:49]: dataclasses are 3.7+ only. On 3.6 (and only there), `setup.py` should require https://pypi.org/project/dataclasses/.

File: pytools/tag.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    """
+    Only one instance of this type of tag may be assigned
+    to a single tagged object.
+    """
+    pass
+
+
+TagsType = FrozenSet[Tag]
+
+# }}}
```
Comment (inducer) [inline pytools/tag.py:91]:
```suggestion
# }}}

# vim: foldmethod=marker
```

File: pytools/tag.py (hunk 2 above)
Comment (nchristensen) [reply pytools/tag.py]: Okay, I think this requires making the forward reference to DottedName a string then.

Comment (nchristensen) [reply pytools/tag.py]: Done, but now the Documentation test is unhappy about something.

File: setup.py [hunk 5]
```diff
@@ -12,6 +13,17 @@
...
+version = sys.version_info
+requirements = [
+          "decorator>=3.2.0",
+          "appdirs>=1.4.0",
+          "six>=1.8.0",
+          "numpy>=1.6.0",
+          ]
+
+if version[0] == 3 and version[1] == 6:
+    requirements.append("dataclasses>=0.7")
```
Comment (inducer) [inline setup.py:25]: I learned literally five minutes ago that [this](https://github.com/inducer/meshmode/pull/59#discussion_r504320314) is a possibility.

File: setup.py [hunk 6]
```diff
@@ -12,6 +13,17 @@
...
+
 setup(name="pytools",
       version=ver_dic["VERSION_TEXT"],
```
Comment (inducer) [inline setup.py:28]: Pytools needs a version bump, and loopy/pytato should declare a dependency on that version.

File: pytools/tag.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+# {{{ docs
+
+__doc__ = """
+.. currentmodule:: pytools
```
Comment (inducer) [inline pytools/tag.py:33]:
```suggestion
```

File: pytools/tag.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+.. autoclass:: Tag
+.. autoclass:: UniqueTag
+
+Supporting Functionality
+------------------------
+
+.. autoclass:: DottedName
+
+.. currentmodule:: pytools.tag
+
```
Comment (inducer) [inline pytools/tag.py:47]:
```suggestion
```

**Review by inducer — COMMENTED** [review]
Thanks for working on this!

File: setup.py (hunk 5 above)
Comment (nchristensen) [reply setup.py:25]: Nice, that is much cleaner!

File: setup.py (hunk 6 above)
Comment (nchristensen) [reply setup.py]: To (2020, 4, 2) or (2020, 5) or (2020, 10)?

Comment (inducer) [reply setup.py]: (2020, 4, 2)

File: pytools/tag.py (hunk 2 above)
Comment (nchristensen) [reply pytools/tag.py]: I'm still not quite certain why the documentation generation fails. I have Sphinx v3.2.1 and 'make html' executes without failing.

File: pytools/tag.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    Should be referenced as :class:`pytools.DottedName`.
+
+.. class:: Tag
+
+    Should be referenced as :class:`pytools.Tag`.
+
+.. class:: UniqueTag
+
+    Should be referenced as :class:`pytools.UniqueTag`.
```
Comment (inducer) [inline pytools/tag.py:62]:
```suggestion
```

File: pytools/tag.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    picklable, and have a reasonably concise :meth:`__repr__`
+    of the form ``dotted.name(attr1=value1, attr2=value2)``.
+    Positional arguments are not allowed.
+
+   .. automethod:: __repr__
+
+   .. note::
+
+       This mirrors the tagging scheme that :mod:`loopy`
+       is headed towards.
```
Comment (inducer) [inline pytools/tag.py:118]:
```suggestion
```

File: pytools/tag.py (hunk 2 above)
Comment (inducer) [reply pytools/tag.py]: You have to run it in "nitpicky mode". See the CI config.

**Comment by inducer** [comment]
Thx!

**Pull request merged**